designs/CentrallyManagingNuGetPackageVersions-VSManagePackages.md

```diff
@@ -0,0 +1,414 @@
+
+## Problem Background
+
+Within Visual Studio, only project systems read and manipulate project files, and provide APIs to other components, such as NuGet. This allows project systems to efficiently handle changes. Although CPVM's `Directory.Packages.props` file is not the project file itself (`csproj`, `vbproj`, and so on), it is imported by the project files and therefore contributes to project evaluation.
+
+To enable NuGet's package management functionality within Visual Studio (Package Manager UI, Package Manager Console, or NuGet's APIs to install packages), NuGet needs agreement with project systems on how APIs should behave with the new feature.
+
+## Who are the customers
+
+All customers managing packages on projects that are using central package version management with projects in Visual Studio IDE.
+
+## Minimum Viable Product Goals
+
+1. When installing a package in a centrally managed project, if the package already exists in the `Directory.Packages.props` file, only the project file is modified. The `PackageReference` element added must not have a version applied.
+2. When installing a package in a centrally managed project, if the package does not already exist in the `Directory.Packages.props` file, then the package version is added to `Directory.Packages.props`, in addition to goal 1 above.
+3. When updating a centrally managed package version, the `Directory.Packages.props` file is updated, and all projects that use the `Directory.Packages.props` must be restored.
+4. When uninstalling/removing a package from a centrally managed project, the `PackageReference` is removed from the project file, but the `PackageVersion` is kept in the `Directory.Packages.props` file.
+
+## Later Phase Goals
+
+5. Removing a `PackageVersion` from the `Directory.Packages.props` file is out of scope for the first version, but will be a requirement in the future.
+
+## Non - Goals
+
+6. This spec is scoped to Visual Studio API interaction between components. Package Manager UI changes are out of scope. Command Line Interface is out of scope.
+7. `Directory.Packages.props` files that import other MSBuild files. If the customer uses this, the behavior is undefined, specifically around which `Directory.Packages.props` file gets updated.
+
+## Scenarios
+
+The following scenarios will be enabled when the proposed design changes will be completed.
+
+### Install a package in a project that already exists in `Directory.Packages.props`
+
+In this scenario, there is a `Directory.Packages.props` file which specifies a `PackageVersion` for `PackageA`. There is a project `Project1`, which does not currently have a reference to `PackageA`. When the customer installs `PackageA` into `Project1`, the project's file is modified to include a `<PackageReference Include="PackageA" />`, which does not specify a version. `Directory.Packages.props` remains unchanged.
+
+#### Before
+
+`$(repo_root)\Directory.Packages.props`
```

Comment:
Do this new API will handle CPVM in MSBuild?

https://github.com/microsoft/MSBuildSdks/tree/master/src/CentralPackageVersions

Reply:
No, the API is for NuGet's CPVM feature.